File: php/reverse.php
#!/usr/bin/php
<?php
/**
 * Reverse a string using only a single char as a temporary variable
 * 
 * See https://github.com/cjwfuller/practicing-c
 * for version that uses NO temporary variable
 */

$test = "some string";
echo $test . "\n";

for( $i = 0; $i <= strlen( $test ) / 2; $i++ ) {
    $tmp = $test[$i];
    $test[$i] = $test[strlen( $test ) - $i - 1];
    $test[strlen( $test ) - $i - 1] = $tmp;
}

echo $test . "\n";
?>

File: c/reverse-string-no-tmp-variable/reverse.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

// Program to reverse a string using the "XOR trick"
// Not an intuitive way of solving this problem but it uses no temp variable
// See https://github.com/cjwfuller/practicing-php for an easier approach
// but with a temp variable
int main() {
    // Input string
    char str[] = "test";
    printf( "Input string: %s\n", str );
    // End of string
    int end = strlen(str) - 1;
    // Start of string
    int start = 0;
    while(start < end) {
        // The "XOR trick"
        str[start] ^= str[end];
        str[end] ^= str[start];
        str[start]^= str[end];
        ++start;
        --end;
    }
    printf( "Reversed string: %s\n", str );
    return 0;
}
